pkg/master/master.go

```diff
@@ -281,183 +294,25 @@ func (m *Master) InstallAPIs(c *Config) {
-		}
-		allGroups = append(allGroups, group)
 	}
 
-	if c.APIResourceConfigSource.AnyResourcesForVersionEnabled(policyapiv1alpha1.SchemeGroupVersion) {
-		policyResources := m.getPolicyResources(c)
-		policyGroupMeta := registered.GroupOrDie(policy.GroupName)
-
-		// Hard code preferred group version to policy/v1alpha1
-		policyGroupMeta.GroupVersion = policyapiv1alpha1.SchemeGroupVersion
-
-		apiGroupInfo := genericapiserver.APIGroupInfo{
-			GroupMeta: *policyGroupMeta,
-			VersionedResourcesStorageMap: map[string]map[string]rest.Storage{
-				"v1alpha1": policyResources,
-			},
-			OptionsExternalVersion: &registered.GroupOrDie(api.GroupName).GroupVersion,
-			Scheme:                 api.Scheme,
-			ParameterCodec:         api.ParameterCodec,
-			NegotiatedSerializer:   api.Codecs,
-		}
-		apiGroupsInfo = append(apiGroupsInfo, apiGroupInfo)
-
-		policyGVForDiscovery := unversioned.GroupVersionForDiscovery{
-			GroupVersion: policyGroupMeta.GroupVersion.String(),
-			Version:      policyGroupMeta.GroupVersion.Version,
-		}
-		group := unversioned.APIGroup{
-			Name:             policyGroupMeta.GroupVersion.Group,
-			Versions:         []unversioned.GroupVersionForDiscovery{policyGVForDiscovery},
-			PreferredVersion: policyGVForDiscovery,
-		}
-		allGroups = append(allGroups, group)
-
+	restOptionsGetter := func(resource unversioned.GroupResource) generic.RESTOptions {
+		return m.GetRESTOptionsOrDie(c, resource)
 	}
 
-	if c.APIResourceConfigSource.AnyResourcesForVersionEnabled(appsapi.SchemeGroupVersion) {
-		appsResources := m.getAppsResources(c)
-		appsGroupMeta := registered.GroupOrDie(apps.GroupName)
-
-		// Hard code preferred group version to apps/v1alpha1
-		appsGroupMeta.GroupVersion = appsapi.SchemeGroupVersion
-
-		apiGroupInfo := genericapiserver.APIGroupInfo{
-			GroupMeta: *appsGroupMeta,
-			VersionedResourcesStorageMap: map[string]map[string]rest.Storage{
-				"v1alpha1": appsResources,
-			},
-			OptionsExternalVersion: &registered.GroupOrDie(api.GroupName).GroupVersion,
-			Scheme:                 api.Scheme,
-			ParameterCodec:         api.ParameterCodec,
-			NegotiatedSerializer:   api.Codecs,
-		}
-		apiGroupsInfo = append(apiGroupsInfo, apiGroupInfo)
-
-		appsGVForDiscovery := unversioned.GroupVersionForDiscovery{
-			GroupVersion: appsGroupMeta.GroupVersion.String(),
-			Version:      appsGroupMeta.GroupVersion.Version,
+	// stabilize order.
+	// TODO find a better way to configure priority of groups
+	for _, group := range sets.StringKeySet(c.RESTStorageProviders).List() {
+		if !c.APIResourceConfigSource.AnyResourcesForGroupEnabled(group) {
+			continue
 		}
-		group := unversioned.APIGroup{
-			Name:             appsGroupMeta.GroupVersion.Group,
-			Versions:         []unversioned.GroupVersionForDiscovery{appsGVForDiscovery},
-			PreferredVersion: appsGVForDiscovery,
+		restStorageBuilder := c.RESTStorageProviders[group]
+		apiGroupInfo, enabled := restStorageBuilder(c.APIResourceConfigSource, restOptionsGetter)
```

Comment:
> Are you using the enabled at all. All the functions I've checked always returned false, or are you missing some additional code pieces still?

Downstream components may choose disable if other conditions are met.  Imagine a resource that manages RCs.  If RCs are disabled, that resource should be disabled, but this logic here can't detect it.  If that resource is the only thing in the group, the group is effectively disabled.